## Wk9 Assignment

## Lincoln Brown

## DSC550-T302

## Professor Werner

In [1]:
# Import the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.pipeline import FeatureUnion
from sklearn.model_selection import GridSearchCV

## Step 1: Import the Dataset

Import the dataset and ensure that it loaded properly

In [2]:
# Import the dataset
loan_df = pd.read_csv('Loan_Train.csv')

# Display the dataframe
loan_df

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

In [3]:
# Print the shape of the original dataframe
print(f"loan_df contains {loan_df.shape[0]} rows and {loan_df.shape[1]} columns.")

loan_df contains 614 rows and 13 columns.


## Step 2: Prepare the Data

Prepare the data for modeling by performing the following steps:
<ul>
    <li>Drop the column “Load_ID.”</li>
    <li>Drop any rows with missing data.</li>
    <li>Convert the categorical features into dummy variables.</li>
</ul>


In [4]:
converted_df = loan_df.copy()
# Drop the Loan_ID column
converted_df = converted_df.drop("Loan_ID", axis=1)

# Label Encode Loan_Status
label_encoder = LabelEncoder()
converted_df['Loan_Status'] = label_encoder.fit_transform(converted_df['Loan_Status'])

# Drop any rows that contain missing values
converted_df = converted_df.dropna(axis=0)

# Convert the categorical features into dummy variables
converted_df = pd.get_dummies(converted_df, drop_first=True)

# Display the converted dataframe
converted_df

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
5               5417             4196.0       267.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History  Loan_Status  Gender_Male  Married_Yes  Dependents_1  \
1               1.0            0         True         True          True   
2               1.0            1         True         True         False   
3               1.0            1         True         True         False   
4               1.0            1         True        False         False   
5               1.0            1         True         True         False   
..              ...          ...          ...          ...           ...   
609             1.0            1        False        False         False   
610             1.0            1         True         True         False   
611             1.0            1         True         True          True   
612             1.0            1         True         True         False   
613             0.0            0        False        False         False   

     Dependents_2  Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
1           False          False                   False              False   
2           False          False                   False               True   
3           False          False                    True              False   
4           False          False                   False              False   
5            True          False                   False               True   
..            ...            ...                     ...                ...   
609         False          False                   False              False   
610         False           True                   False              False   
611         False          False                   False              False   
612          True          False                   False              False   
613         False          False                   False               True   

     Property_Area_Semiurban  Property_Area_Urban  
1                      False                False  
2                      False                 True  
3                      False                 True  
4                      False                 True  
5                      False                 True  
..                       ...                  ...  
609                    False                False  
610                    False                False  
611                    False                 True  
612                    False                 True  
613                     True                False  

[480 rows x 15 columns]

In [5]:
# Print the shape of the dataframe
print(f"The converted_df has {converted_df.shape[0]} rows and {converted_df.shape[1]} columns.")

The converted_df has 480 rows and 15 columns.


## Step 3: Split the Data

Split the data into a training and test set, where the “Loan_Status” column is the target.


In [6]:
# Separate features from the target feature
X = converted_df.drop("Loan_Status", axis=1)
y = converted_df["Loan_Status"]

# Create train and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Step 4: Create a Pipeline

Create a pipeline with a min-max scaler and a KNN classifier (see section 15.3 in the Machine Learning with Python Cookbook).


In [7]:
# Create a MinMax Scaler
minmax = MinMaxScaler()

# Create a KNN Classifier
knn = KNeighborsClassifier(n_neighbors=5, n_jobs=-1)

# Create a pipeline
pipe = Pipeline([("scaler", minmax), ("classifier", knn)])

## Step 5: Fit a KNN Classifier

Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set. Note: Fitting a pipeline model works just like fitting a regular model.

In [8]:
# Fit the KNN classifier to the data with the pipeline
pipe.fit(X_train, y_train)

# Report the accuracy
accuracy = pipe.score(X_test, y_test) 

print(f"The pipeline achieved an accuracy score of {accuracy}.")

The pipeline achieved an accuracy score of 0.78125.


## Step 6: Create a Search Space

Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10. (see section 15.3 in the Machine Learning with Python Cookbook).


In [9]:
# Create a Search Space
range_space = range(1,11)
search_space = [{"classifier__n_neighbors": range_space}]


## Step 7: Fit a Grid Search
# Create a grid search

Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.


In [10]:
# Create a grid search

classifier = GridSearchCV(pipe, search_space, cv=5, verbose=0).fit(X_train, y_train)

print(f"Best Value for n_neighbor parameter: {classifier.best_params_['classifier__n_neighbors']}")

Best Value for n_neighbor parameter: 3


## Step 8: Find the Best Model Score on Test Set

Find the accuracy of the grid search best model on the test set. Note: It is possible that this will not be an improvement over the default model, but likely it will be.


In [11]:
test_classifier_score = classifier.best_estimator_.score(X_test, y_test)
# Print the best model's score on the test set
print(f"Accuracy of the best model on the test set: {test_classifier_score}")


Accuracy of the best model on the test set: 0.7916666666666666


## Step 9: Include Logistic Regression and Random Forest Models

Now, repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the hyperparameter values in section 12.3 of the Machine Learning with Python Cookbook.


In [12]:
# Expand the search space to include logistic regression and random forest models

search_space_expanded = [{'classifier': [LogisticRegression(max_iter=500, solver='liblinear')],
                          'classifier__penalty': ['l1', 'l2'], 
                          'classifier__C': np.logspace(0, 4, 10)},
                         {'classifier': [RandomForestClassifier()],
                          'classifier__n_estimators': [10, 100, 1000],
                          'classifier__max_features': [1, 2, 3]}]


In [13]:
# Create a Grid Search with the expanded pipeline
gridsearch = GridSearchCV(pipe, search_space_expanded, cv=5, verbose=0)

# Fit the model
best_model = gridsearch.fit(X_train, y_train)

In [14]:
# Best Estimator
best_model.best_estimator_
print(f"The best model hyperparameters were: {best_model.best_estimator_['classifier']}")

The best model hyperparameters were: LogisticRegression(C=7.742636826811269, max_iter=500, penalty='l1',
                   solver='liblinear')


In [15]:
# Best Parameters
best_model.best_params_

{'classifier': LogisticRegression(max_iter=500, solver='liblinear'),
 'classifier__C': 7.742636826811269,
 'classifier__penalty': 'l1'}

In [16]:
# Print Best Model Accuracy Score
test_best_model_score = best_model.best_estimator_.score(X_test, y_test)
print(f"Best Model Accuracy Score: {test_best_model_score}")

Best Model Accuracy Score: 0.8229166666666666


## Conclusion

From the outputs above, we can see that hyperparameter tuning achieved a higher accuracy score of .82 versus the original .78 using a knn classifier and still better than the tuned knn classifier model which achieved an accuracy score of .79. 

At first, I was achieving lower scores after hyperparameter tuning and realized that I had not included drop_first in my pd.get_dummies function. This was leaving the original columns that had been encoded and was affecting my model's performance. 

The model that achieved the best performance was LogisticRegression with a C value of 7.742636826811269, max_iter of 500, and a penalty for l1 errors, using the liblinear solver. 

This exercise helped me understand the importance of hyperparameter tuning and that without the use of tools like GridSearchCV, it would be very tedious to manually find the best hyperparameters. I am looking forward to working with more advanced modules in the future like Hyperopt or Optuna. 